AX: AXTextMarker::findMarker, a frequently called function, can copy all the text-runs strings, causing performance issues on pages with lots of text

twilco opened the pull request:
#### d05070c1f5a99cbb7b38243b5924899c1c7280bc
<pre>
AX: AXTextMarker::findMarker, a frequently called function, can copy all the text-runs strings, causing performance issues on pages with lots of text
<a href="https://bugs.webkit.org/show_bug.cgi?id=294996">https://bugs.webkit.org/show_bug.cgi?id=294996</a>
<a href="https://rdar.apple.com/154332688">rdar://154332688</a>

Reviewed by Joshua Hoffman.

For groups of text runs that contain non-ASCII characters, AXTextMarker::findMarker needs to use CachedTextBreakIterator
to determine how many offsets to move forward or backward. This is relevant for things like multi-byte emojis. CachedTextBreakIterator
takes a StringView, but we were always doing the far less efficient thing of combining the text from all AXTextRun structs
into a single String, requiring multiple allocations along the way. This is a major performance problem on webpages that
have lots of text (e.g. 40 seconds of a 60 second sample was spent building and allocating strings underneath findMarker).

We had to do it this way because CachedTextBreakIterator needs the full string from all runs, but because each AXTextRun
owned its own string, we couldn&apos;t just naively use StringView here, because you can&apos;t combine disparate StringViews (one
for each AXTextRun) into one larger StringView (that&apos;s not how StringViews work).

Solve the issue by making AXTextRuns own a singular string for all runs, and add start and end index fields to AXTextRun
that point into this string. This allows us to use StringViews in many more places, which this commit does.

* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::streamTextRuns):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::toString const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::findParagraph const):
(WebCore::AXTextMarker::findWordOrSentence const):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::debugDescription const):
(WebCore::AXTextRuns::localRect const):
(WebCore::AXTextRuns::substring const): Deleted.
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::AXTextRun):
(WebCore::AXTextRun::length const):
(WebCore::AXTextRuns::AXTextRuns):
(WebCore::AXTextRuns::at const):
(WebCore::AXTextRuns::operator[] const):
(WebCore::AXTextRuns::runLength const):
(WebCore::AXTextRuns::lastRunLength const):
(WebCore::AXTextRuns::lineID const):
(WebCore::AXTextRuns::toString const):
(WebCore::AXTextRuns::toStringView const):
(WebCore::AXTextRuns::substring const):
(WebCore::AXTextRuns::runString const):
(WebCore::AXTextRuns::runStringView const):
(WebCore::AXTextRuns::runStartsWithLineBreak const):
(WebCore::AXTextRuns::runEndsWithLineBreak const):
(WebCore::AXTextRun::debugDescription const): Deleted.
(WebCore::AXTextRun::startsWithLineBreak const): Deleted.
(WebCore::AXTextRun::endsWithLineBreak const): Deleted.
(WebCore::AXTextRun::characterAt const): Deleted.
(WebCore::AXTextRun::textLength const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):

Canonical link: <a href="https://commits.webkit.org/296681@main">https://commits.webkit.org/296681@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aff5d989f4cf0cb8496fa78b1cb1d4155873afb7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109171 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19256 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114378 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59449 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111134 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29510 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37424 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82963 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112119 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23484 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98328 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63412 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22874 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16471 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59027 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92851 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16513 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/rotated-cat-off-top-edge.html imported/w3c/web-platform-tests/feature-policy/feature-policy-frame-policy-timing.https.sub.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117493 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36215 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26800 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91977 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36587 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91783 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23395 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36706 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14454 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32064 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36111 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41606 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35797 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39131 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37488 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth reviewed (approved):
lgtm!

I think we talked about comparing stringValue to the concatenated textRuns strings to assert that they're equal, and maybe even avoid saving both - that'd be even easier now.

twilco committed:
AX: AXTextMarker::findMarker, a frequently called function, can copy all the text-runs strings, causing performance issues on pages with lots of text
https://bugs.webkit.org/show_bug.cgi?id=294996
rdar://154332688

Reviewed by Joshua Hoffman.

For groups of text runs that contain non-ASCII characters, AXTextMarker::findMarker needs to use CachedTextBreakIterator
to determine how many offsets to move forward or backward. This is relevant for things like multi-byte emojis. CachedTextBreakIterator
takes a StringView, but we were always doing the far less efficient thing of combining the text from all AXTextRun structs
into a single String, requiring multiple allocations along the way. This is a major performance problem on webpages that
have lots of text (e.g. 40 seconds of a 60 second sample was spent building and allocating strings underneath findMarker).

We had to do it this way because CachedTextBreakIterator needs the full string from all runs, but because each AXTextRun
owned its own string, we couldn't just naively use StringView here, because you can't combine disparate StringViews (one
for each AXTextRun) into one larger StringView (that's not how StringViews work).

Solve the issue by making AXTextRuns own a singular string for all runs, and add start and end index fields to AXTextRun
that point into this string. This allows us to use StringViews in many more places, which this commit does.

* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::streamTextRuns):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::toString const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::findParagraph const):
(WebCore::AXTextMarker::findWordOrSentence const):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::debugDescription const):
(WebCore::AXTextRuns::localRect const):
(WebCore::AXTextRuns::substring const): Deleted.
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::AXTextRun):
(WebCore::AXTextRun::length const):
(WebCore::AXTextRuns::AXTextRuns):
(WebCore::AXTextRuns::at const):
(WebCore::AXTextRuns::operator[] const):
(WebCore::AXTextRuns::runLength const):
(WebCore::AXTextRuns::lastRunLength const):
(WebCore::AXTextRuns::lineID const):
(WebCore::AXTextRuns::toString const):
(WebCore::AXTextRuns::toStringView const):
(WebCore::AXTextRuns::substring const):
(WebCore::AXTextRuns::runString const):
(WebCore::AXTextRuns::runStringView const):
(WebCore::AXTextRuns::runStartsWithLineBreak const):
(WebCore::AXTextRuns::runEndsWithLineBreak const):
(WebCore::AXTextRun::debugDescription const): Deleted.
(WebCore::AXTextRun::startsWithLineBreak const): Deleted.
(WebCore::AXTextRun::endsWithLineBreak const): Deleted.
(WebCore::AXTextRun::characterAt const): Deleted.
(WebCore::AXTextRun::textLength const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):

Canonical link: https://commits.webkit.org/296681@main
